internal/chooserbenchmark/visualization.go

```diff
@@ -0,0 +1,330 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package chooserbenchmark
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	tabWidth               = 4
+	separator              = "#"
+	bar                    = "*"
+	displayWidth           = 120
+	serverHistogramBuckets = 10
+	latencyHeight          = 15
+	latencyWidth           = 101
+)
+
+type ClientGroupMeta struct {
+	name      string
+	count     int
+	rps       int
+	reqCount  int
+	resCount  int
+	histogram []int
+}
+
+type ServerGroupMeta struct {
+	name         string
+	servers      []*Server
+	histogram    []int
+	requestCount int
+	meanLatency  time.Duration
+}
+
+type DisplayConfig struct {
+	idLen             int
+	freqLen           int
+	counterLen        int
+	tableStarUnit     float64
+	histogramStarUnit float64
+	latencyStarUnit   float64
+}
+
+type Visualizer struct {
+	clientGroupNames []string
+	serverGroupNames []string
+	serverData       map[string]*ServerGroupMeta
+	clientData       map[string]*ClientGroupMeta
+	buckets          []int
+
+	config *DisplayConfig
+}
+
+func NewServerGroupMeta(groupName string, servers []*Server, buckets []int) *ServerGroupMeta {
+	requestCount := 0
+	counters := make([]int, len(servers))
+	for i, server := range servers {
+		counter := server.counter
+		requestCount += counter
+		counters[i] = counter
+	}
+	sort.Ints(counters)
+
+	histogram := make([]int, serverHistogramBuckets)
+	idx := 0
+	for _, counter := range counters {
+		for buckets[idx] < counter {
+			idx++
+		}
+		histogram[idx]++
+	}
+
+	return &ServerGroupMeta{
+		name:         groupName,
+		servers:      servers,
+		histogram:    histogram,
+		requestCount: requestCount,
+		meanLatency:  servers[0].latency.median,
+	}
+}
+
+func NewClientGroupMeta(groupName string, clients []*Client) *ClientGroupMeta {
+	rps := int(float64(time.Second) / clients[0].mu)
+	reqCount, resCount := 0, 0
+	histogram := make([]int, BucketLen)
+	for _, client := range clients {
+		reqCount += client.reqCounter
+		resCount += int(client.resCounter.Load())
+		for i, freq := range client.histogram {
+			histogram[i] += int(freq.Load())
+		}
+	}
+	return &ClientGroupMeta{
+		name:      groupName,
+		count:     len(clients),
+		rps:       rps,
+		reqCount:  reqCount,
+		resCount:  resCount,
+		histogram: histogram,
+	}
+}
+
+func NewVisulizer(ctx *Context) *Visualizer {
```

Comment:
this function is _very_ long. Let's add comments and maybe try to break this up into smaller functions